server/src/main/java/de/coronavirus/imis/api/EnumDataController.java

```diff
@@ -0,0 +1,132 @@
+package de.coronavirus.imis.api;
+
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.Trigger;
+import org.springframework.scheduling.TriggerContext;
+
+@RestController
+@RequestMapping("/api/enum-data")
+@RequiredArgsConstructor
+public class EnumDataController {
+
+  private final TaskScheduler taskScheduler;
+  private final ObjectMapper jsonMapper;
+  private final ClientHttpRequestFactory requestFactory;
+
+  private static final long HI_COMPANIES_CACHE_LIFETIME_MILLIS = 60 * 60 * 1000;
+  private List<String> healthInsuranceCompanies;
+
+  @Autowired
+  public EnumDataController(TaskScheduler taskScheduler, ObjectMapper jsonMapper) {
+    this.taskScheduler = taskScheduler;
+    this.jsonMapper = jsonMapper;
+
+    SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+    requestFactory.setConnectTimeout(7000);
+    requestFactory.setReadTimeout(5000);
+    this.requestFactory = requestFactory;
+
+    FetchHealthInsuranceCompanies fhicTask = new FetchHealthInsuranceCompanies();
+    this.taskScheduler.schedule(fhicTask, fhicTask.new TaskTrigger());
+  }
+
+  @GetMapping("/health-insurance-companies")
+  public List<String> getHealthInsuranceCompanies(
+    @RequestParam(required = false) String search,
+    @RequestParam(required = false) Optional<Integer> count) {
+
+    Stream<String> result;
+    if (this.healthInsuranceCompanies != null) {
+      result = this.healthInsuranceCompanies.parallelStream();
+
+      if (search != null) {
+        result = result.filter((var companyName) -> companyName.contains(search));
+      }
+
+      if (count.isPresent()) {
+        result = result.limit(count.get());
+      }
+    } else {
+      result = Arrays.stream(new String[0]);
+    }
+
+    return result.collect(Collectors.toList());
+  }
+
+  public class FetchHealthInsuranceCompanies implements Runnable {
+    private boolean lastRunSuccessful;
+
+    public void run() {
+      try {
+        // Fetch recent data
+        ClientHttpResponse companiesResponse = EnumDataController.this.requestFactory.createRequest(
```

Comment:
Ich wäre extrem vorsichtig 3rd party APIs ohne Zustimmung zu benutzen

Reply:
Aus deren agb:
https://www.gesetzlichekrankenkassen.de/impressum/agb/agb.html
Der Benutzer ist berechtigt, die Leistungen des Online-Angebots, sowie dessen Inhalte ausschließlich für private, nicht-kommerzielle Zwecke zu verwenden. Die Informationen dürfen nur in begrenztem Umfang ausschließlich an Privatpersonen zu privaten Zwecken weitergegeben werden. Jegliche Nutzung der Informationen zu geschäftlichen Zwecken, insbesondere die gewerbliche Nutzung, die Veröffentlichung in Medien, im Internet oder anderen Publikationsformen, ist ohne vorherige Zustimmung der Kassensuche GmbH unzulässig.

Reply:
Autsch, da hast du natürlich vollkommen Recht :man_facepalming:. Die Namen der deutschen Krankenkassen sollten aber grundsätzlich als Fakten in einem statischen Array auf dem IMIS-Server speicherbar sein, oder?

Reply:
Die Abhängigkeit von der externen API habe ich nun durch eine lokale Ressource (Datei) des IMIS-Webservers ersetzt. Die darin gespeicherte Liste erhebt keinerlei Anspruch auf Vollständigkeit :see_no_evil: